providers/implementations/rands/crngt.c

```diff
@@ -137,34 +156,41 @@ size_t prov_crngt_get_entropy(PROV_DRBG *drbg,
                                OSSL_SELF_TEST_DESC_RNG);
     }
 
-    while ((q = rand_pool_bytes_needed(pool, 1)) > 0 && attempts-- > 0) {
-        s = q > sizeof(buf) ? sizeof(buf) : q;
-        if (!crngt_get_entropy(libctx, crngt_glob->crngt_pool, buf, md, &sz))
+    for (t = bytes_needed; t > 0;) {
+        /* Care needs to be taken to avoid overrunning the buffer */
+        s = t >= CRNGT_BUFSIZ ? CRNGT_BUFSIZ : t;
+        entbuf = t >= CRNGT_BUFSIZ ? entp : buf;
+        if (!crngt_get_entropy(drbg->provctx, crngt_glob->md, entbuf, md, &sz))
             goto err;
+        if (t < CRNGT_BUFSIZ)
+            memcpy(entp, buf, t);
         /* Force a failure here if the callback returns 1 */
         if (OSSL_SELF_TEST_oncorrupt_byte(st, md))
             memcpy(md, crngt_glob->crngt_prev, sz);
         if (!prov_crngt_compare_previous(crngt_glob->crngt_prev, md, sz)) {
             crng_test_pass = 0;
             goto err;
         }
-        if (!rand_pool_add(pool, buf, s, s * 8))
-            goto err;
+        /* Update for next block */
         memcpy(crngt_glob->crngt_prev, md, sz);
-        t += s;
-        attempts++;
+        entp += s;
+        t -= s;
     }
-    r = t;
-    *pout = rand_pool_detach(pool);
-err:
+    r = bytes_needed;
+    *pout = ent;
+    ent = NULL;
+
+ err:
     OSSL_SELF_TEST_onend(st, crng_test_pass);
     OSSL_SELF_TEST_free(st);
-    OPENSSL_cleanse(buf, sizeof(buf));
```

Comment:
Don't we still need to do this if we used the `buf` in the last iteration of the loop above?

Reply:
Yes, we do need this.  `buf` came back to avoid the overflow.
I've added it.